### ISSUE TITLE
New unified advanced installation dialog for all mod installers

### ISSUE THREAD

**Issue by Al12rs** [ModOrganizer2/modorganizer#1031]
The installer dialog should:

- Primarily get the name of the new mod.
- Offer possible names and sensible alternatives.
- Show which mods are already installed with the same names/same Nexus IDs (same name is important for mods without ID info).
- Allow the user to either merge or replace one of the listed existing mods.
- Have an advanced section to set mod metadata:
    - Priority (maybe choose separator to send to)
    - Enabled status
    - Nexus ID
    - Categories
    - Notes

We currently have these installers:
- [ ] Quick
- [ ] Manual
- [ ] Fomod
- [ ] NCC Fomod
- [ ] Bain

**Comment by AnyOldName3** [comment]
The OMOD installer would go a lot faster if someone else did all the dialogues. I'm not sure the features here capture everything OMOD can do, though. Many of the fields we could use to autofill/suggest stuff are listed here: http://timeslip.chorrol.com/obmmm/index.html.

The main options dialogue would be sensible to reuse, too. The features OMODs allow for each decision are:
* A list of names/titles.
* Whether they're tickboxes or radio buttons.
* Optionally, which are selected by default.
* Optionally, a long description, which may be in RTF format (as in the Microsoft one that old Windows UI stuff supported natively, and not what Qt calls the subset of HTML you can use in labels). If we were having a general-purpose MO2 dialogue, though, forcing installers to covert things to HTML would be reasonable.
* Optionally, a list of images (erri120's OMOD framework uses paths, but it would be reasonable to demand they're converted to QImage or QPixmap or whatever).

Before I got busy, I had something that could do most of this okay-ish, but only okay-ish.

**Comment by Al12rs** [comment]
@AnyOldName3 The dialog is supposed to be used before the actual installation procedure, if the user lets things as default the installer plugin can still set all these things.
Like for Fomods, I would show this dialog first and then procede with the normal Fomod choice dialogs.

**Comment by Al12rs** [comment]
Maybe copy this over to the Omod installer issue

**Comment by Holt59** [comment]
> The dialog is supposed to be used before the actual installation procedure, if the user lets things as default the installer plugin can still set all these things.

I think this is kind of incompatible with:

> - Allow the user to either merge or replace one of the listed existing mods. 

I don't think users want to choose to merge or replace before knowing what is inside the mod (installed or to be installed). I personally find this already very annoying with the NCC installer.

We probably need two dialogs: one before the actual installer and one after. The one before could be given to the installer so that they can re-use it instead of showing a new one: you basically want a dialog with a top-header with mod name / whatever, a middle widget that installer can customize, and a bottom button row that installer can also customize.

**Comment by Al12rs** [comment]
That sounds quite complicated though and would add to the amount of clicks someone that just wants to install a new mod without tweaking anything needs to press.
Maybe add a button to show a preview of the contents (visual tree like we offer for bsa contents) so that people can check out with a click from that dialog. Assuming there is still the Manual install button and cancelling the manual setup brings back this dialog then users can just use that to have a look at the contents instead (but preview download contents would be a cool right click menu on the downloads tab).

**Comment by Holt59** [comment]
> That sounds quite complicated though and would add to the amount of clicks someone that just wants to install a new mod without tweaking anything needs to press.

I don't see what it would add compare to the current state of things?

> Maybe add a button to show a preview of the contents (visual tree like we offer for bsa contents) so that people can check out with a click from that dialog. 

Showing the contents make no sense for FOMOD or BAIN installer, or even other installers unless the archive is perfectly formatted. So if we want to show any kind of "conflict" with the existing mod (so that user can decide on merging / replacing / renaming), it's impossible to do it before the actual installation procedure.

**Comment by AnyOldName3** [comment]
None of the data in an OMOD is in the top level. There are several layers of zipping. You're never going to get a nice tree out of it.

Also, we can't really get scripted installers to go inside a parent widget. I don't know about C# FOMODs, but C# OMODs can and do create native windows as and when they please.

**Comment by Holt59** [comment]
> Also, we can't really get scripted installers to go inside a parent widget. I don't know about C# FOMODs, but C# OMODs can and do create native windows as and when they please.

Yes, those are kind of differents. But currently you still need a pre-dialog to get the name of the mod, etc. 

The OMOD installer is external, so this is hard to apply because the user will be asked to replace / merge / replace even before starting, which I personally find not very user-friendly, but that's one the custom installers work currently.
